src/analysis.rs

```diff
@@ -0,0 +1,547 @@
+use syntax::hir::{
+    Hir, HirKind, Literal, ClassBytes, ClassBytesRange,
+    Class, Visitor, RepetitionRange, RepetitionKind
+};
+use syntax::hir;
+use utf8_ranges::Utf8Sequences;
+
+/// True iff the given expression is one-pass
+pub fn is_onepass(expr: &Hir) -> bool {
+    hir::visit(expr, IsOnePassVisitor::new()).unwrap()
+}
+
+struct IsOnePassVisitor(bool);
+
+impl Visitor for IsOnePassVisitor {
+    type Output = bool;
+    type Err = ();
+
+    fn finish(self) -> Result<bool, ()> {
+        Ok(self.0)
+    }
+
+    fn visit_pre(&mut self, hir: &Hir) -> Result<(), ()> {
+        if !self.0 {
+            return Ok(())
+        }
+
+        match hir.kind() {
+            &HirKind::Concat(ref es) => self.check_concat(&es),
+            &HirKind::Alternation(ref es) => self.check_alternation(&es),
+            &HirKind::Repetition(ref rep) => {
+                if fset_of(&*rep.hir).is_empty() {
+                    self.0 = false;
+                }
+            }
+            &HirKind::Class(ref cls) => self.check_cls(cls),
+            _ => ()
+        }
+
+        Ok(())
+    }
+}
+
+impl IsOnePassVisitor {
+    fn new() -> Self {
+        IsOnePassVisitor(true)
+    }
+
+    fn check_concat(&mut self, es: &[Hir]) {
+        let mut empty_run = vec![];
+
+        for e in NestedConcat::new(es) {
+            // TODO(ethan):yakshaving factor the determination of when
+            //     a regex accepts_empty out into a separate function,
+            //     so that we don't compute the whole first set when we
+            //     don't need to.
```

Comment:
Have you seen? https://docs.rs/regex-syntax/0.6.2/regex_syntax/hir/struct.Hir.html#method.is_match_empty

(That's an example of a property that is computed inductively as a HIR is constructed.)